src/MessagePack.Annotations/Attributes.cs

```diff
@@ -109,10 +110,13 @@ public UnionAttribute(int key, Type subType)
         /// </summary>
         /// <param name="key">The distinguishing value that identifies a particular subtype.</param>
         /// <param name="subType">The full name (should be assembly qualified) of the derived or implementing type.</param>
+#if NET8_0_OR_GREATER
+        [RequiresUnreferencedCode("The type named in 'subType' may be trimmed.")]
+#endif
         public UnionAttribute(int key, string subType)
```

Comment:
You should be able to mark the `subType` with a DynamicallyAccessedMembers attribute, which will tell the trimmer that the string is a `Type`. And so any Type that is passed in on this string will be preserved.

I'm not sure on exactly which `DynamicallyAccessedMemberTypes` to use, it may work with just `.None`, or you may need to add `PublicParameterlessConstructor`. @sbomer and @agocke would know for sure. `PublicParameterlessConstructor` would be the safest to ensure it is kept.

```suggestion
        public UnionAttribute(int key, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] string subType)
```

This would allow you to remove the `RequiresUnreferencedCode` attribute on this constructor.

Reply:
`None` didn't work. But `PublicParameterlessConstructor` worked. There's no guarantee that such a constructor is declared on this type, so I hope it works either way.